File: src/agent.php
<?php
require_once(dirname(__FILE__).'/../vendor/autoload.php');
use Symfony\Component\Process\Process;
use Symfony\Component\Process\PhpProcess;
use League\CLImate\CLImate;
use Parse\ParseClient;
use Parse\ParseObject;
use Parse\ParseException;
use Parse\ParseQuery;

$config = require_once(dirname(__FILE__).'/config.php');

$climate = new CLImate;
$arguments = [
    'filter' => [
        'prefix'        => 'f',
        'longPrefix'    => 'filter',
        'description'   => 'Filter which filetype that will be monitored',
        'defaultValue'  => '*'
    ],
    'interval' => [
        'prefix'        => 'i',
        'longPrefix'    => 'interval',
        'description'   => 'Interval between creating new record for changes',
        'defaultValue'  => '120' // seconds (2 minutes by default)
    ],
    'name'  => [
        'prefix'        => 'n',
        'longPrefix'    => 'name',
        'description'   => 'This session\'s name',
        'defaultValue'  => 'lorem ipsum'
    ],
    'directory' => [
        'prefix'        => 'd',
        'longPrefix'    => 'directory',
        'description'   => 'Which folder to monitor',
        'defaultValue'  => '*'
    ]
];
$climate->arguments->add($arguments);
try{
    $climate->arguments->parse();
}catch(Exception $e){
    $climate->out('you need this yo!');
    exit;
}

$interval_limit = $climate->arguments->get('interval');
$monitor_path =  str_replace('\\', '/', $climate->arguments->get('directory'));
// Include filter into the folder's path
// TODO: check if the value is not a directory
$filters = explode(' ', $climate->arguments->get('filter'));
$monitor_pattern = sprintf('%s/**/%s', $monitor_path, implode('||', $filters));

$format = json_encode(['"filename"' => '"%FILENAME%"', '"event"' => '"%FSEVENT%"']);
$cmd = sprintf('filewatcher "%s" "echo %s"', $monitor_pattern, $format);
echo $cmd;
$process = new Process($cmd);
$process->start();
$climate->red()->out('MONITORING STARTED');

ParseClient::initialize($config['parse']['app_id'], 
                        $config['parse']['rest_key'], 
                        $config['parse']['master_key']);

if(function_exists('posix_getpwuid'))
{
  $processUser = posix_getpwuid(posix_geteuid());
  $user = $processUser['name'];
}
else
{
  $user = getenv('username');
}

// Create an identifier for this machine
$id_path = dirname(__FILE__).'/.codemon';
$machineID = '';
if(file_exists($id_path))
{
  $machineID = file_get_contents($id_path);
}

if(!file_exists($id_path) or !$machineID)
{
  $factory = new RandomLib\Factory;
  $generator = $factory->getMediumStrengthGenerator();
  $machineID = $generator->generateString(32);
  file_put_contents($id_path, $machineID); 
}
$climate->out('machineID: ' . $machineID);

$monObj = ParseObject::create('MonSession');
$monObj->set('user', $user);
$monObj->set('machineID', $machineID);
$monObj->set('directory', $monitor_path);
$monObj->set('name', $climate->arguments->get('name'));
echo $monObj->save();
$mon_session_id = $monObj->getObjectId();
$climate->out('session created: ' . $mon_session_id);
    
while($process->isRunning()){
    if($output = $process->getIncrementalOutput())
    {
        $json = json_decode($output,true);
        if(!$json){ continue; }
        
        $filetype = filetype($json['filename']);
        
        // Find if there are similar changes saved
        $query = new ParseQuery('FileChanges');
        $query->equalTo('parent', $monObj);
        $query->equalTo('filename', $json['filename']);
        $query->equalTo('type', $filetype);
        $query->descending('updatedAt');
        $results = $query->find();
        
        $flag_new = 1;
        if(count($results) > 0)
        {
            $file_changes = $results[0];
            // Compare the duration
            $updatedAt = $file_changes->getUpdatedAt();
            $now = new DateTime('now');
            $interval = $now->diff($updatedAt);
            $minutes = $interval->format('%i');
            $seconds = $minutes * 60;
            $climate->out(sprintf('%s minutes ago updated', $minutes));
            if($seconds < $interval_limit)
            {
                $climate->out('Still within interval limit');
                $flag_new = 0;
            }
            else
            {
                $flag_new = 1;
            }
        }
       
        if($flag_new)
        {
            $climate->out('creating new record for changes');
            $file_changes = new ParseObject('FileChanges');
            $file_changes->set('filename',$json['filename']);
            $file_changes->set('type', $filetype);
            $file_changes->set('parent', $monObj);
            $climate->out('changes saved');
        }
        else
        {
            $climate->out('updating record for changes');
        }
        
        $file_changes->set('event',$json['event']);
        $file_changes->set('content', ($filetype === 'file' && $json['event'] != 'deleted') ? file_get_contents($json['filename']):'');
        $response = $file_changes->save();
        echo $output;
        echo $response;
    }
}

File: src/agent.py
import argparse
import pdb
import os
import sys
import time
import getpass
import yaml
import random
import string
from watchdog.observers import Observer  
from watchdog.events import PatternMatchingEventHandler
from parse_rest.connection import register
from parse_rest.datatypes import Object
from parse_rest.query import QueryResourceDoesNotExist

class CodeMonHandler(PatternMatchingEventHandler):
    patterns = []
    ignore_patterns = []
    ignore_directories = False
    case_sensitive = False
    fileChanges = ''
    
    def __init__(self, args=None, session=None):
        self.patterns = args.filters
        self.ignore_patterns = ['*.git*']
        self.session = session
        self.args = args

    def process(self, event):
        """
        event.event_type 
            'modified' | 'created' | 'moved' | 'deleted'
        event.is_directory
            True | False
        event.src_path
            path/to/observed/file
        """
        # the file will be processed there
        # print(event.src_path, event.event_type)  # print now only for debug
        
        within_interval = self.within_interval(event.src_path) # check if file is within interval, if it is update existing record
        #self.fileChanges.filename = os.path.basename(event.src_path)
        self.fileChanges.filename = os.path.abspath(event.src_path)
        self.fileChanges.type = 'directory' if event.is_directory else 'file'
        self.fileChanges.parent = self.session.as_pointer
        self.fileChanges.event = event.event_type
        if event.event_type == 'modified' or (event.event_type == 'created' and event.is_directory == False):
            print('Saving %s file contents: %s' % (event.event_type, self.fileChanges.filename))
            with open(event.src_path, 'r') as contents:
                self.fileChanges.content = contents.read()
        else:
            self.fileChanges.content = ''
            
        self.fileChanges.save()
          
        
    def within_interval(self, src_path):
        fileChangesObj = Object.factory('FileChanges')
        try:
          fileChanges = fileChangesObj.Query.all().filter(parent = self.session.as_pointer,
                                   filename = os.path.basename(src_path),
                                   type = 'file').limit(1)
          #pdb.set_trace()
          fileChanges = fileChanges[0] if len(fileChanges) > 0 else False
        except QueryResourceDoesNotExist:
          fileChanges = False
          
        
        if fileChanges:
            current_time = time.gmtime()
            time_diff = (time.mktime(current_time) - time.mktime(fileChanges.updatedAt.timetuple()))
            if time_diff > self.args.interval_limit:
                print('Over interval limit')
                self.fileChanges = fileChangesObj()
                return False
            else:
                print('within interval')
                self.fileChanges = fileChanges
                
        else:
            self.fileChanges = fileChangesObj()
        
        return True
        

    def on_modified(self, event):
        self.process(event)

    def on_created(self, event):
        self.process(event)
        
if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Monitor current folder')
    parser.add_argument('-f', '--filter', dest='filters', nargs='*',
                        default=['*.py', '*.txt'], help='Filter which filetype that will be monitored, default is all')
    parser.add_argument('-i', '--interval', dest='interval_limit', default=120,
                        help='Interval between creating a new record for changes')
    parser.add_argument('-n', '--name', dest='monitor_name', 
                        default='%s - %s' % (getpass.getuser(), os.path.basename(os.getcwd())), help='This session\'s name')
    parser.add_argument('-d','--directory', dest='monitor_path', 
                        default='.', help='Which folder to monitor')
    args = parser.parse_args()
    
    f = open('%s/config.yaml' % os.path.abspath(os.path.dirname(sys.argv[0])))          
    config = yaml.safe_load(f)
    f.close()
    register(config['parse']['app_id'], config['parse']['rest_key'], master_key=None)
    
    # Create session
    monObj = Object.factory('MonSession')
    monObj = monObj()
    monObj.user = getpass.getuser()
    monObj.machineID = ''.join(random.choice(string.ascii_uppercase + string.digits + string.ascii_lowercase) for _ in range(32))
    monObj.directory = os.path.abspath(args.monitor_path)
    monObj.name = args.monitor_name
    monObj.save()
    print('Session created: %s' % monObj.objectId)
    
    observer = Observer()
    observer.schedule(CodeMonHandler(args=args,session=monObj), path=args.monitor_path, recursive=True)
    observer.start()

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        observer.stop()

    observer.join()
    